### PR TITLE
Initiate market dashboard 

### PR DESCRIPTION
**Pull request by songyoskeng**
Initiate new widget for market dashboard
with dashboard grid style and dark theme color.

Next Station,
I will create example for market and display graph with [d3.js](https://www.d3-graph-gallery.com/)

### CODE REVIEW

**Comment by songyoskeng** [comment]
Oh, I think I implemented wrong file. Let me fix

**Comment by songyoskeng** [comment]
I've already connect example_market_dashboard and display message from socket (widget:alerts channel).

**Comment by thananon** [comment]
This is a good start. I can see a new widget with some layout but I cannot get any information on the widget. I don't want this dashboard to be the exact design to what I posted in #23 . 

Please feel free to continue working on it. I will not merge a work-in-progress PR.

**Comment by thananon** [comment]
Can we have some screenshot to show the current state of this PR?

**Comment by songyoskeng** [comment]
@thananon 
sorry about late reply 😢.

This's current progress.

image


I think It's ok for market dashboard v.1.0 👍 

This dashboard can display
- current transaction
- latest winners
- richest players
- current session income & session payout

If you think it's ready for launch it now, pls remove WIP and  review the code. 🙏 

for next improvement 💯 
- more previous data for session income & session payout.
- investigate to init market dashboard data and improve loading state.
- more display card (that the bot can provide)

p.s. i never test with the real bot maybe has a little bug in 9armbot.js 😭 .

**Comment by kaozaza2** [comment]
i think should use native javascript function instead of lodash for more performance.

**Comment by songyoskeng** [comment]
@kaozaza2 

No problem 😉 

if you guys concern about the performance, I'll improve it later.

**Comment by kaozaza2** [comment]
add more dependencies is break my storage haha 😂

**Comment by songyoskeng** [comment]
@kaozaza2 

I've already remove `lodash` from package.json 🙌 

Still using in `show.ejs` via cdn, Could I use it?

**Comment by kaozaza2** [comment]
Thank a lot.

File: webapp/example_market_dashboard.js [hunk 1]
```diff
@@ -0,0 +1,184 @@
+/**
+ * Test send message every 3s and display GIF
+ */
+
+var dayjs = require('dayjs')
+const webapp = require('.')
+const _ = require('lodash')
```
Comment (kaozaza2) [inline webapp/example_market_dashboard.js:7]: lodash dependency is removed.

Comment (songyoskeng) [reply webapp/example_market_dashboard.js:7]: OH  😢

Comment (songyoskeng) [reply webapp/example_market_dashboard.js]: fixed

File: 9armbot/9armbot.js [hunk 2]
```diff
@@ -172,6 +217,73 @@ function gacha(channel, user, amount) {
...
+            webapp.socket.io().emit("widget::market_dashboard", {
+                key: MARKET_KEY.LATEST_WINNERS,
+                data: {
+                    winners: gachaWinners
+                }
+            });
+        }
+        
+        const currentTopCoinPlayers = player.getCoinTop(MIN_TOP_COIN_PLAYER)
+        if(currentTopCoinPlayers.length > 0 && gachaRateType !== GACHA_RATE_TYPE.SALT){
```
Comment (kaozaza2) [inline 9armbot/9armbot.js:278]: `getCoinTop` function is undefined, you forgot add this in `player.js` ?

**Review by kaozaza2 — CHANGES_REQUESTED** [review]

File: 9armbot/9armbot.js (hunk 2 above)
Comment (kaozaza2) [reply 9armbot/9armbot.js:278]: Oops, wrong line comment.

Comment (songyoskeng) [reply 9armbot/9armbot.js:278]: Hmm? 
`player` refers to variable at line 27 in `9armbot.js` that is used  many places in the file. 

If everything is ok, Please resolve this conversation.

Comment (kaozaza2) [reply 9armbot/9armbot.js:278]: Sorry about this, I forgot fully check in player class

**Review by kaozaza2 — APPROVED** [review]

File: 9armbot/9armbot.js (hunk 2 above)
Comment (songyoskeng) [reply 9armbot/9armbot.js:278]: No problem

File: webapp/widgets/market_dashboard/views/show.ejs [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+                $(`#txn-${txns[0].timestamp}`).remove()
+                txns.shift()
+              },300)
+            }
+            txns.push(data)
+            let message = displayGachaTransaction(username, amount, gain, rate, txnTime)
+            $("#transaction").append(`<li id="txn-${timestamp}" class="animate__animated animate__fadeInUp" >${message}</li>`)
+          } else if (key === "LATEST_WINNERS"){
+            const {winners} = data
+            if(_.isEqual(previousWinners, winners)) return
```
Comment (kaozaza2) [inline webapp/widgets/market_dashboard/views/show.ejs:317]: I found lodash again at line 317, 334, 434

**Review by kaozaza2 — CHANGES_REQUESTED** [review]

File: webapp/widgets/market_dashboard/views/show.ejs (hunk 3 above)
Comment (kaozaza2) [reply webapp/widgets/market_dashboard/views/show.ejs:317]: sorry again, i think it use from module forgot about frontend 😢🙏,

**Review by kaozaza2 — APPROVED** [review]

File: webapp/widgets/market_dashboard/views/show.ejs (hunk 3 above)
Comment (songyoskeng) [reply webapp/widgets/market_dashboard/views/show.ejs:317]: No worries 😸

**Comment by songyoskeng** [comment]
@thananon Can I merge this PR ?

**Comment by songyoskeng** [comment]
**don't forget `npm install` first** (install [dayjs](https://day.js.org/) package)

Run with this command for mock data and see how dashboard  work.
```bash
node webapp/example_market_dashboard.js
```

after run this click [Link to preview](http://localhost:3000/widget/market_dashboard/preview)

I will write this process on `readme.md` next time.

**Comment by thananon** [comment]
will give a try run today.

**Pull request closed without merging**